[NO MERGE] BIP331 Ancestor Package Relay

glozow opened the pull request:
**WORK IN PROGRESS.** Please do not run it for any use other than testing.
  
This PR is not meant for merge. This branch exists to help reviewers see what package relay would look like all together. I will open separate PRs for these components and expect to add more tests, docs, and polish along the way. This PR contains all of the functionality built in a linear manner. However, since there are pieces in various areas of the codebase and they can make progress in parallel, commits don't necessarily need to be merged in this order.

**See #27463 for what PR(s) are open for review right now.**

## Note to Reviewers

The major purpose of this PR is to solicit "approach" review.

This is a large project, and the first few p2p commits essentially define the interface. I'd like to get rough consensus on approach before we start looking at code details and merging PRs, because I believe it will help us "get useful stuff in" faster and avoid premature optimizations.

Here are some questions I hope are answered before we try to merge anything (originally https://github.com/bitcoin/bitcoin/pull/27742#issuecomment-1572205313):

1. Does the proposed protocol look sound?
2. Are these 3 milestones appropriate?
3. Is there important functionality that is in the "todo improvements" section but should be included in one of the 3 milestones? Alternatively, is there not-that-important stuff in the milestones that we should defer until later?
4. Does it make sense to have this PeerManager <-> TxDownloadMan interface?

Comments about naming, typos, code details, etc. are also appreciated but please note I may wait until their respective PRs are open to incorporate your comments. Thank you for your patience.

### Design Docs

BIP: https://github.com/bitcoin/bips/pull/1382
Orphanage changes: https://gist.github.com/glozow/7c0ff639f996e660146314edffa6f06c

## Project Structure

### 3 Major Milestones

This project is split into 3 milestones, each of which gives us something useful.

1. Modularize and improve orphan-handling (also some refactoring).
	- Introduce a `TxDownloadManager` class, responsible for downloading transactions that have been announced to us.
	- Use all announcers as potential candidates for resolving an orphan. Add a `TxRequestTracker` Orphan Request Tracker which helps track orphans we need to resolve. Preferentially request orphan resolution from outbounds, preferred relay, etc., over inbounds.

2. When package relay peers are available, use `ancpkginfo` instead of missing parents when handling orphans.
	- Add `sendpackages` negotiation logic.
	- Add a separate rejections filter for transactions that are eligible for reconsideration when validated together as a package, so that children of low-feerate transactions are still considered.
	- Send `getdata(MSG_ANCPKGINFO)` to package relay peers for orphan resolution. Use `ancpkginfo` to request missing ancestors through normal individual transaction relay.

3. Download and validate ancestor packages using `getpkgtxns` and `pkgtxns`.
	- Add support for `getpkgtxns` and `pkgtxns`. Send a `pkgtxns` using the list of missing transactions from `ancpkginfo`.
	- Ensure we can process "normal" orphans even if a peer is trying to overwhelm/churn our orphanage. Do this by "opportunistically" protecting orphans from random eviction if they were sent by package relay peers, and redownloading orphans if we cannot afford to protect them. Each peer is allocated a certain amount of orphans they can protect at a time ("token bucket" style but the number of tokens is static for now). Outbound peers get more than inbounds.
	- If a transaction's parent(s) are below the fee filter, don't announce it (save the bandwidth of legacy nodes).

###  Todo improvements 
These could be added to the milestones or deferred until basic functionality is deployed.
- Store orphans serialized instead of as CTransactionRefs to significantly reduce their memory usage.
- Perhaps try to keep orphans in disk and/or process them asynchronously, given the incredibly DoSable nature of orphan handling.
- Dynamically allocate tokens for orphan protection. For example, if a long-standing inbound peer continuously provides good packages for orphans, they should have more tokens. If a peer is obviously serving us garbage, reduce their tokens.
- Detect when we have received all the transactions for a package, regardless of how (individual or block or other), and return `PackageToValidate`  in `GetTxToReconsider`.
- New format for mempool.dat, packages instead of transactions.

glozow committed:
[p2p] add tx package tracker

glozow committed:
[rpc] expose package relay on rpc

glozow committed:
[p2p] respond to getdata(ancpkginfo) requests with ancpkginfo

Conditions for responding to an ancpkginfo request are identical to
responding to a tx. That is, we only provide ancpkginfo if we would have
provided tx data as well.

glozow committed:
[validation/p2p] separate TxValidationResult and rejects filter for low fees

We wouldn't want to add a low-feerate transaction to m_recent_rejects
because usually that means we won't give the child a chance
(`fRejectedParents`); we instead want to fetch this orphan's low-feerate
parents and potentially accept them.  However, we also shouldn't
revalidate transactions/packages that have already been rejected.
Continue to cache low-fee rejections, but separately.

Note: when ephemeral anchor transaction fails due to a missing spender,
we should similarly put it in the reconsiderable filter.

glozow committed:
MOVEONLY: move package checks into helper functions

glozow committed:
[packages] AncestorPackage sorts and builds ancestor subsets

We cannot require that peers send topologically sorted lists, because we
cannot check for this property without ensuring we have the same chain
tip and ensuring we have the full ancestor set. Instead, add the ability
to handle arbitrarily ordered transaction lists.

glozow committed:
[test util] multi-input multi-output CreateValidMempoolTransaction

glozow committed:
[p2p] respond to getpkgtxns with pkgtxns or notfound MSG_PKGTXNS

glozow committed:
[refactor] wrap TxOrphanage inside TxPackageTracker

glozow committed:
[refactor] use txpackagetracker for orphan resolution

No behavior change in this commit.

glozow committed:
[p2p] use all orphan announcers as potential parent sources

Makes orphan handling more robust. We will also use the orphan
resolution tracker for requesting ancpkginfo. The timeout in
p2p_segwit.py has been increased to account for the fact that parents
are no longer immediately requested.

glozow committed:
[p2p] Resolve orphans by requesting ancpkginfo and requesting txdata

For now, each tx in ancpkginfo is treated like an individual tx
announcement. When the tx arrives, it is validated individually as well.

glozow committed:
[p2p] signal support for package relay

glozow committed:
[validation] skip and pre-fill txns with irreversibly invalid ancestors

Packageifier calculates the in-package ancestors. We already know
this tx will fail because it depends on something invalid (specifically
for a non-policy reason). We also know that it is missing at least one
input (the tx that did not and will not make it into our mempool).

Note to reviewers: slight behavior change. If this tx has multiple
errors, there may be a difference in which one is returned. For
example, if the tx has a dust output in addition to relying on
the invalid tx, we will return TX_MISSING_INPUTS when we previously
would have returned TX_NOT_STANDARD. This is simply because dust is
checked earlier within PreChecks().

Add a test that we don't quit *too* early and reject transactions we
should keep.

glozow committed:
[validation] call AcceptSingleTransaction when only 1 package tx left

Avoid calling PackageMempoolChecks() when there is only 1 transaction.
Note to reviewers: there is a slight change in the error type returned,
as shown in the txpackage_tests change.  When a transaction is the last
one left in the package and its fee is too low, this returns a PCKG_TX
instead of PCKG_POLICY. This interface is clearer;
"package-fee-too-low" for 1 transaction would be a bit misleading.

glozow committed:
[validation] validate packages by submitting each tx's ancestor sub-packages

This results in the incentive-compatible transactions ending up in our
mempool. Prior to this commit, if parents within the package relied on
each other, we could end up (1) accepting a low-feerate child or (2)
rejecting high-feerate parents. Instead of validating each transaction
*individually* in turn, validate each one with their in-package ancestor
set. This means parents with inter-dependencies are validated correctly,
while we continue using aggregate totals for package feerate.

glozow committed:
[validation] skip last tx in AcceptPackage "submit individually" loop

Prior to this commit, if the last tx is invalid for a policy, missing
inputs, or "mempool full", we validate it twice (once inside this loop,
and then again afterwards). This is unnecessary.
If we come to the last transaction in the package, break.

This needs to be after the !subpackage quit early condition so that
that the child is present in the results map.

subfill the child state for missing inputs

glozow committed:
[txorphanage] add ability to protect from random eviction

glozow committed:
[txpackagetracker] track packages to download

glozow committed:
[p2p] use ancpkginfo to make getpkgtxns requests and validate packages

glozow committed:
[txpackagetracker] handle pkgtxns/notfound, return validation work items

glozow committed:
[packages] hash multiple transactions

glozow committed:
scripted-diff: rename CheckPackage to IsPackageWellFormed

-BEGIN VERIFY SCRIPT-
sed -i 's/CheckPackage(/IsPackageWellFormed(/g' $(git grep -l CheckPackage)
-END VERIFY SCRIPT-

glozow committed:
[txpackagetracker] add orphan resolution tracker

glozow committed:
[txpackagetracker] create ancpkginfo requests using orphan resolution tracker

glozow committed:
[txorphange] use wtxid instead of txid, handle nullptr

glozow committed:
[fuzz] AncestorPackage

glozow committed:
[policy] allow any ancestor package, not just child-with-unconfirmed-parents

We can safely allow any ancestor package since Packageifier can handle
things that are out of order. Remove the check that "all unconfirmed
parents are present" because, even if a tx is missing inputs, the other
transactions may be worth validating.

glozow committed:
[txorphanage] return erased wtxids from EraseForBlock

glozow committed:
-- Part 1: Orphan Resolution Module --

Create TxPackageTracker module, responsible for handling orphan
resolution. Enable node to retry orphan parent downloading from multiple
peers who announced the tx. Use preferred peers and avoid overloading
peers with too many requests.

Making the orphanage more reliable is also necessary because it becomes
part of the "critical path" for some transactions.  A 0-fee parent
bumped by a child *must* be relayed via package relay; if the orphanage
overflows or is churned by a malicious peer, the package would be
censored from the network.

glozow committed:
-- Part 2: Negotiate Package Relay, Request Ancestor Package Info For Orphans --

glozow committed:
-- Part 3: Download and Validate Packages --

glozow committed:
[unit test] txpackagetracker orphan tracking

glozow committed:
[unit test] txpackagetracker tests

glozow committed:
[txpackagetracker] negotiation logic

glozow committed:
[unit test] PkgInfoAllowed, ReceivedAncPkgInfo, ReceivedPkgTxns

glozow committed:
[log] add category TXPACKAGES for orphanage and package relay

glozow committed:
[txrequest] GetCandidatePeers and ResetRequestTimeout

glozow committed:
[txorphanage] impose a maximum total size of orphans

No effect for now.
TODO: add per-peer limits on orphan memory usage

glozow committed:
[p2p] don't inv fee-bumping children to non-pkgrelay peers

These transactions can be more common with package relay, but
non-package-relay peers will only waste bandwidth and orphanage space,
then end up not accepting them. Save their bandwidth by dropping
announcements of transactions that have below-fee-filter parents.

glozow committed:
[rpc] allow submitpackage to be called outside of regtest

glozow committed:
[txorphanage] track bytes provided per peer

glozow committed:
[txpackagetracker] delete unused OrphanageAddTx

Adding/deleting is now strictly handled by TxPackageTracker, and should
not be exposed to the public.

glozow committed:
[txorphanage] support multiple announcers

glozow committed:
[unit test] orphanage: multiple announcers, eviction

glozow committed:
[doc] no notfound handling for orphan resolution by parent txid

Note this is status quo.

glozow committed:
[functional test] orphan handling with multiple announcers and dosers

glozow committed:
[unit test] orphanage protection

glozow committed:
[txpackagetracker] track protections and token bucket it

glozow committed:
[doc] add release note for submitpackage

glozow committed:
[validation] require child feerate > package feerate

glozow committed:
[unit test] parent pay for child is not allowed

Co-authored-by: Andrew Chow <achow101@gmail.com>

glozow committed:
MOVEONLY: rename AcceptPackage to AcceptAncestorPackage

glozow committed:
[functional test] package relay works despite peers DoSing orphanage

instagibbs commented:
> Approach review is very welcome on this PR

didn't real whole OP until now. In my not expert opinion, the orphan handling seems reasonable, and I think it makes to open it ASAP to take it out of draft

glozow commented:
Offline feedback suggested I clarify what I mean by "approach feedback welcome" before "I open separate PRs."

This is a large project, and the first few p2p commits essentially define the interface. I'd like to get rough consensus on approach before we start looking at code details and merging PRs, because I believe it will help us "get useful stuff in" faster and avoid premature optimizations.

Here are some approach questions I think we should answer before diving in:
1. Does the proposed protocol look sound?
2. Are these 3 milestones appropriate?
3. Is there important functionality that is in the "todo improvements" section but should be included in one of the 3 milestones? Alternatively, is there not-that-important stuff in the milestones that we should defer until later?
4. Does it make sense to have this `PeerManager` <-> `TxPackageTracker` <-> `TxOrphanage` relationship?

instagibbs commented:
> redownloading orphans if we cannot afford to protect them

Is the idea here that protection is a bandiwdth optimization to avoid fetching an orphan twice in a row? Reasonable if so, just want this to be explicitly stated if so.

glozow commented:
> > redownloading orphans if we cannot afford to protect them
> 
> Is the idea here that protection is a bandiwdth optimization to avoid fetching an orphan twice in a row in the presence of malicious peers or random churn? Reasonable if so, just want this to be explicitly stated if so.

Yes exactly. Worse is we want to avoid a situation where we requested the rest of the package while the orphan was in our orphanage, but then it fell out while we were waiting for the rest of the transactions. Then, we should re-download the orphan, but what do we do with the rest of them? We basically have to redownload the whole thing again, otherwise we might end up in an endless cycle of redownloading if we put them into an orphanage but don't necessarily protect them from eviction.

Here is a more detailed doc for the orphanage protection and "token bucket" scheme: https://gist.github.com/glozow/7c0ff639f996e660146314edffa6f06c.

instagibbs commented on src/txorphanage.cpp:
```diff
@@ -74,21 +99,43 @@ int TxOrphanage::_EraseTx(const uint256& txid)
-        auto it_last = m_orphan_list.back();
-        m_orphan_list[old_pos] = it_last;
-        it_last->second.list_pos = old_pos;
+    if (it->second.list_pos >= 0) {
```
Should this function ever be called with a protected orphan? Tests don't cover these lines for protected peers(added an assert, never hit), and seems suspect that it would delete things below this block if it's protected.

glozow replied:
> Should this function ever be called with a protected orphan?

Yes. It's called when orphans expire as well.

> Tests don't cover these lines for protected peers

Sounds like I neglected to write a test for this. Expiry is pretty long, but theoretically we can hit this if we have multiple peers stalling us on the resolution of this orphan. We might happen to be trying package relay with one of the peers when the expiration is reached.

instagibbs replied:
Ah! Didn't put 2 and 2 together. Comment-worthy

instagibbs reviewed (commented):
Along with the gist, focused on orphan protection and bucketing for concept understanding. Dumping comments related to those, and left a note on the gist.

instagibbs reviewed (commented):
Along with the gist, focused on orphan protection and bucketing for concept understanding. Dumping comments related to those, and left a note on the gist.

glozow commented on src/txorphanage.cpp:
```diff
@@ -188,7 +191,7 @@ void TxOrphanage::LimitOrphans(unsigned int max_orphans)
     }
     FastRandomContext rng;
-    while (m_orphans.size() > max_orphans || m_total_orphan_bytes > MAX_ORPHAN_TOTAL_SIZE)
+    while (m_orphan_list.size() > max_orphans || m_total_orphan_bytes > MAX_ORPHAN_TOTAL_SIZE)
```
Ah yes, that's a bug! This `while` condition is very strange since one part applies to the unprotected orphanage and the other applies to the total orphanage. I think we should change the second condition to be on `m_total_unprotected_orphan_bytes`, yes.

glozow replied:
Fixed here

glozow commented on src/txorphanage.cpp:
```diff
@@ -176,9 +207,13 @@ void TxOrphanage::AddChildrenToWorkSet(const CTransaction& tx)
             for (const auto& elem : it_by_prev->second) {
+                Assume(elem->second.announcers.size() >= 1);
+                if (elem->second.announcers.empty()) break;
+                // Pick the first peer from announcers set.
```
Basically arbitrary. The relevance is basically which peer's message processing is blocked to process this, and would get a misbehaving score if the tx is invalid. Anyone who announced it is equally deserving of this burden imo. I suppose a more fair way to pick is uniformly randomly.

ariard commented on src/node/txpackagetracker.h:
```diff
@@ -58,6 +59,36 @@ class TxPackageTracker {
+    void AddOrphanTx(NodeId nodeid, const uint256& wtxid, const CTransactionRef& tx, bool is_preferred, std::chrono::microseconds reqtime);
+
+    /** Number of packages we are working on with this peer. Includes any entries in the orphan
+     * tracker, in-flight orphan parent requests (1 per orphan regardless of how many missing
```
If I understand correctly the rate-limiting accounting introduced here, whatever the number of orphan parent missing, there will be an in-flight orphan parent requests issued (i.e if we’re missing 10 parents, we’re going to issue 10 in-flight orphan parent requests).

I think this should be documented as multi-party applications and contracting protocols transactions issuers might have to assemble their packages in function.

ariard replied:
> So from a time-sensitive contract perspective, things should be strictly better than today?

I think this is a different concern I’m aiming to underscore. It’s not related about “buckets”-capture by the attackers, rather than if you have a package of {A, B, C, D} with E as a CPFP and we have rate-limiting based on the way the parents are fetched (i.e individiually rather than in batch) the equivalent package of A+E will propagate better.

So if you would like A to confirm as soon as you can because it’s a 0-conf dual-funding, you wanna probably go for A+E.

ariard replied:
Okay so I’ve been looking into the code, and yes to my understanding there is _no_ processing penalty at `GETPKGTXNS` reception in `ProcessMessage` than our child has 1 or N parents, we’re calling `FindTxForGetdata` to compose the response `PKGTXNS`. This invalidate this specific source of concern.

> we have rate-limiting based on the way the parents are fetched (i.e individiually rather than in batch) the equivalent package of A+E will propagate better.

I’ll maintain there is still an interpedency between package rate-limits and package compositions, though telling what to do a LN node is unclear and can be deferred to ulterior documentation.

ariard commented on src/node/txpackagetracker.h:
```diff
@@ -58,6 +59,36 @@ class TxPackageTracker {
+     * Should be called when tx is added to mempool.
+     * Should not be called when a tx fails validation.
+     * */
+    void FinalizeTransactions(const std::set<uint256>& valid, const std::set<uint256>& invalid);
```
This is unclear what level of enforcement is granted to a “final” decision, if the transactions are added to any transaction-relay download filters such as `m_recent_rejects` or `m_recently_announced_invs`. Otherwise, in case of blocks re-orgs dropping out the first package component and a second package component still stuck in filters, this can be source of package propagation failure.

glozow replied:
Note that the rejection filters are cleared every time the chain tip changes, so I don't think we can get failures this way

ariard replied:
> Note that the rejection filters are cleared every time the chain tip changes, so I don't think we can get failures this way

Yes, I think we should be careful to clean the whole package state in all filters at every block change to avoid “half” state interfering. `m_recent_rejects_reconsiderable_were_rejected` say “reset this filter when the chain tip changes”, of course it includes _backward_ change like re-orgs.

ariard commented on src/txorphanage.h:
```diff
@@ -43,7 +43,8 @@ class TxOrphanage {
     int EraseTx(const uint256& wtxid) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
-    /** Erase all orphans announced by a peer (eg, after that peer disconnects) */
+    /** Maybe erase all orphans announced by a peer (eg, after that peer disconnects). If an orphan
```
I think there can be a source of concerns if all the announced orphans are second components from a single common parent. All our peers would announce those second-stage components, only one of them will succeed in the mempool validation and other will be refused as conflicts, I think.

instagibbs replied:
Can you elaborate on this concern? The orphan that isn't received yet may still be in orphanage, and will be connected to its parents normally?

ariard replied:
> Can you elaborate on this concern? The orphan that isn't received yet may still be in orphanage, and will be connected to its parents normally

So if you have 3 peers and they announce A, B, C all with different wtxids while they’re spending the same parent txid, do we have a concern of the parent being fetched from multiple peers, or download should be dedup when we receive the `ancpkginfo`s ?

instagibbs replied:
> download should be dedup when we receive the ancpkginfos ?

If it's in your orphanage it won't be fetched again. See `ReceivedAncPkgInfo`

instagibbs commented on src/txorphanage.h:
```diff
@@ -72,6 +72,13 @@ class TxOrphanage {
         LOCK(m_mutex);
         return m_total_orphan_bytes;
     }
+    size_t BytesFromPeer(NodeId peer) const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
```
Orphan usage would effect the CPFP transactions, IIUC, not the commitment transactions. The CPFP is deemed high enough, sent to the peer, the peer holds the child tx in orphan set and requests the full package, which isn't size-bound(beyond current mempool limits).

ariard replied:
Just to be sure we’re in sync on `BytesFromPeer` and `MAX_ORPHANAGE_USAGE`, it’s correct this a per-peer limit on the space in vbytes occupied by received orphans ?

If yes, I think this still introduce an interdependency between the number of outgoing transactions bytes and the maximum flow of LN commitment transactions and this limit should be documented (a la `doc/policy/`), or even better announced as part of `sendpackages` info.

instagibbs reviewed (commented):
another bucket of comments to respond to later

instagibbs commented on src/net_processing.cpp:
```diff
@@ -3992,6 +4255,78 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        }
+        std::vector<std::pair<uint256, bool>> txdata_status;
+        txdata_status.reserve(package_wtxids.size());
+        std::vector<uint256> pkgtxns_to_request;
```
this is never read from, just written to

instagibbs commented on src/net_processing.cpp:
```diff
@@ -3992,6 +4255,78 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        std::vector<uint256> pkgtxns_to_request;
+        for (const auto& wtxid : package_wtxids) {
+            AddKnownTx(*peer, wtxid);
+            if (AlreadyHaveTx(GenTxid::Wtxid(wtxid), /*include_orphanage=*/false)) {
```
what's the important of not checking orphanage here, then checking the orphanage inside `ReceivedAncPkgInfo`?

glozow commented on src/net_processing.cpp:
```diff
@@ -3319,6 +3336,16 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         if (greatest_common_version >= WTXID_RELAY_VERSION) {
             m_connman.PushMessage(&pfrom, msg_maker.Make(NetMsgType::WTXIDRELAY));
+            if (m_enable_package_relay) {
```
I think it's weird to set `-blocksonly=1 -packagerelay=1`, but also not the end of the world. In that case we'd want to not send "sendpackages." Does that answer the question?

glozow commented on src/protocol.cpp:
```diff
@@ -223,6 +226,6 @@ std::vector<std::string> serviceFlagsToStr(uint64_t flags)
 {
-    assert(inv.IsGenTxMsg());
+    assert(inv.IsGenTxMsg() || inv.IsMsgAncPkgInfo());
     return inv.IsMsgWtx() ? GenTxid::Wtxid(inv.hash) : GenTxid::Txid(inv.hash);
```
Ooh good catch!

glozow commented on src/net_processing.cpp:
```diff
@@ -817,6 +847,33 @@ class PeerManagerImpl final : public PeerManager
+     * Memory used: 1.3 MB
+     * FIXME: this filter can probably be smaller, but how much smaller?
+     */
+    CRollingBloomFilter m_recent_rejects_reconsiderable GUARDED_BY(::cs_main){120'000, 0.000'001};
```
Good catch 👍

ariard commented on src/txorphanage.h:
```diff
@@ -50,7 +50,12 @@ class TxOrphanage {
     std::vector<uint256> EraseForBlock(const CBlock& block) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
-    /** Limit the orphanage to the given maximum */
+    /** Limit the orphanage to the given maximum. Delete orphans whose expiry has been reached.
```
I think the expiry orphans could be announced at the protocol-level, as part of the “package-link”, i.e somewhere in `sendpackages` or in `ancpkinfo`. Those limits could be announced from the peer’s `PeerManagerImpl` with a push-interface (e.g the `ZMQNotificationInterface` ), that way a Lightning node could adapt its package broadcasting strategy accordingly.

ariard commented on src/txorphanage.h:
```diff
@@ -65,6 +70,16 @@ class TxOrphanage {
+    /** Protect an orphan from eviction from the orphanage getting full. The orphan may still be
+     * removed due to expiry. If the orphan is already protected (by any peer), nothing happens.
+     */
+    void ProtectOrphan(const uint256& wtxid) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```
Should we implement blessing of all the orphans of a peer with “relay” permissions ? I can definitely see Lightning deployment where you have an inner Bitcoin Core node and multiple external ones, where you load-balance your flows to avoid transaction-relay fingerprinting based on obvious patterns like HTLC-preimage linked to off-chain channel XYZ.

mzumsande reviewed (commented):
I reviewed part 1 (orphan handling), will continue to the p2p part next week.

These changes make sense to me on a high level, I think it would be ok to open a separate non-draft PR for this - 
 maybe after adding some more context to commit messages, for example it would make review easier if there is a short explanation why something is changed, even if it becomes apparent in later commits.

glozow commented on src/init.cpp:
```diff
@@ -492,6 +493,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-i2psam=<ip:port>", "I2P SAM proxy to reach I2P peers and accept I2P connections (default: none)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-i2pacceptincoming", strprintf("Whether to accept inbound I2P connections (default: %i). Ignored if -i2psam is not set. Listening for inbound I2P connections is done through the SAM proxy, not by binding to a local address and port.", DEFAULT_I2P_ACCEPT_INCOMING), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-onlynet=<net>", "Make automatic outbound connections only to network <net> (" + Join(GetNetworkNames(), ", ") + "). Inbound and manual connections are not affected by this option. It can be specified multiple times to allow multiple networks.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-packagerelay", strprintf("[EXPERIMENTAL] Support relaying transaction packages (default: %u)", node::DEFAULT_ENABLE_PACKAGE_RELAY), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
I think mentioning specific policy here would make things more confusing tbh. I can add a mention to BIP331?

ariard replied:
Yes adding a mention in BIP331 than a p2p packages can support multiple versions of policy regimes e.g nversion=3. 

For flexibility towards the node operators in their resource management (especially for DoS protection on low-performance hosts), I think we should have a `acceptnversion=3`setting, that way a node operator can opted in package relay though not in the most computationally expensive type of policy regimes. The utility of such setting is theoretical as long as we have one policy regime for p2p packages, like right now.

glozow committed:
fixup p2p: require PKG_RELAY_PKGTXNS negotiated for getpkgtxns

glozow committed:
fixup: swap PKG_RELAY bits

glozow commented:
Thanks everyone for the reviews!
Last push:
- Implemented changes that were discussed on the BIP PR https://github.com/bitcoin/bips/pull/1382
    - Maximum of 100 transactions per `getpkgtxns` request
    - Changed `versions` field in `sendpackages` to be 64b instead of 32b.
    - Added a `PKG_RELAY_PKGTXNS` bit for negotiating `getpkgtxns`/`pkgtxns` support in `sendpackages`.
- Fixed some of the bigger problems I found / pointed out by reviewers, e.g.
    - https://github.com/bitcoin/bitcoin/pull/27742#discussion_r1229835172
    - https://github.com/bitcoin/bitcoin/pull/27742#discussion_r1224428421
    - https://github.com/bitcoin/bitcoin/pull/27742#discussion_r1213639103
- Rebased on master for the conflicts

Given the positive approach feedback, I'm going to polish up the Milestone 1 branch (incorporating comments from here as well) and open a non-draft PR for it.

glozow commented on src/txorphanage.h:
```diff
@@ -79,6 +80,10 @@ class TxOrphanage {
 
+    /** Remove a peer from an orphan's announcers list, erasing the orphan if this is the only peer
+     * who announced it. If the orphan doesn't exist or does not list this peer as an announcer, do nothing. */
+    void EraseOrphanOfPeer(const uint256& wtxid, NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```
Removed, thanks (other PR)

glozow reviewed (commented):
Opened #28031 for milestone 1, addressing the comments here pertaining to those commits.

glozow commented on src/txorphanage.h:
```diff
@@ -24,7 +24,7 @@ static constexpr size_t MAX_ORPHAN_TOTAL_SIZE{100 * MAX_STANDARD_TX_WEIGHT};
 class TxOrphanage {
 public:
-    /** Add a new orphan transaction */
+    /** Add a new orphan transaction. If the tx already exists, add this peer to its list of announcers. */
```
Done (in other PR)

mzumsande commented on src/net_processing.cpp:
```diff
@@ -4078,38 +4095,20 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-                }
-
-                if (m_txpackagetracker->OrphanageAddTx(ptx, pfrom.GetId())) {
+                if (!had_before && m_txpackagetracker->OrphanageHaveTx(GenTxid::Wtxid(ptx->GetWitnessHash()))) {
```
Thanks for the explanation, makes sense. I got confused there somehow.

mzumsande commented on src/txorphanage.h:
```diff
@@ -61,7 +65,16 @@ class TxOrphanage {
+        return m_total_orphan_bytes;
+    }
 protected:
+    size_t m_total_orphan_bytes{0};
```
Hmm, I think the maximum possible value is currently `101 * MAX_STANDARD_TX_WEIGHT` (after that, `LimitOrphans()` would kick in), so an `unsigned int` should suffice in theory? Maybe use a `uint64_t` if this was to be made configurable by users (see comment above, but not sure if that makes sense).

ariard commented on src/net_processing.cpp:
```diff
@@ -3012,14 +3170,86 @@ bool PeerManagerImpl::ProcessOrphanTx(Peer& peer)
+    LOCK(cs_main);
+    // We won't re-validate the exact same transaction or package again.
+    if (m_recent_rejects_reconsiderable.contains(GetPackageHash(package.m_unvalidated_txns))) {
+        // Should we do anything else here?
```
So `m_recent_rejects_reconsiderable` logs transaction on its wtxid (which is good to avoid witness inflation changing the result) however isn’t that two restrictive if you have PKG1 submitted with A+B, A is too low with B and then you have A+B’ where A is good enough ? Or is that `TX_LOW_FEE` when the transaction isn’t good enough to meet “mempool min fee not meet” / “min relay fee not met” ?

If it’s correct, I would still favor to exempt package transaction from min relay fee checks as it’s some awful interdependency for pre-signed lightning transaction :)

glozow replied:
Yes, that's the idea of "reconsiderable." If you get A again with a different package, you reconsider it. If you get A+B again, you don't. In this set of changes, `TX_LOW_FEE` is used to decide to put something in the reconsiderable cache.

ariard replied:
Okay “reconsiderable” makes sense to me, still have to review it’s correctly implemented. 

> If it’s correct, I would still favor to exempt package transaction from min relay fee checks as it’s some awful interdependency for pre-signed lightning transaction :)

For this concern, in fact it’s already addressed by the latest state of bip 331 code and nversion=3 documentation, confusion is mine, I thought it was still present from previous package relay reviews.

ariard commented on src/validation.cpp:
```diff
@@ -1423,39 +1438,74 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
+            if (single_res_it != subpackage_result.m_tx_results.end()) {
+                const auto single_res = single_res_it->second;
+                if (single_res.m_state.GetResult() != TxValidationResult::TX_MEMPOOL_POLICY &&
+                    single_res.m_state.GetResult() != TxValidationResult::TX_LOW_FEE &&
```
I think this check is correct in what it aims to achieve, i.e not apply the two mempool min fee checks on ancestor package transaction, if it received as such from `AcceptAncestorPackage` / `ProcessNewPackage`, however this is unclear if this exemption apply for all BIP331 package, indifferently of their transactions versions, or only for nversion=3 ones.

For context, see https://github.com/lightningdevkit/rust-lightning/pull/2415#discussion_r1263189013

mzumsande commented on src/node/txpackagetracker.cpp:
```diff
@@ -195,36 +213,45 @@ class TxPackageTracker::Impl {
+                Assume(info_per_peer.find(nodeid) != info_per_peer.end());
+                // Add the orphan's wtxid as-is.
+                LogPrint(BCLog::TXPACKAGES, "\nResolving orphan %s, requesting by ancpkginfo from peer=%d\n", gtxid.GetHash().ToString(), nodeid);
+                results.emplace_back(gtxid);
```
I think that the splitting into commits is not ideal here: In ae2664793b1d0d86bfa32c3e635d83b1ee083e60, we add logic to return the wtxid, expecting it to be requested by `MSG_ANCPKGINFO`. But the net_processing code to actually do this is only added in dd6fb139b8887f3400a65dc0d4c4ee54cf53d0bc, so that at ae2664793b1d0d86bfa32c3e635d83b1ee083e60, we'd send out a `MSG_TX` getdata, thus re-requesting a TX we already know. These two things should go together into one commit.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3440,6 +3457,16 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             m_connman.PushMessage(&pfrom, msg_maker.Make(NetMsgType::WTXIDRELAY));
+            if (m_enable_package_relay) {
+                m_txpackagetracker->ReceivedVersion(peer->m_id);
+                if (!m_ignore_incoming_txs) {
```
We should also disable sending "sendpackages" for outgoing block-relay-only connections and probably also feelers, would suggest to just use `RejectIncomingTxs()`.
Also, I think this should be moved up and combined with the `m_enable_package_relay` check two lines above, because I see no reason to call `m_txpackagetracker->ReceivedVersion()` in these scenarios.

mzumsande reviewed (commented):
Some initial comments on part 2 (Negotiate Package Relay, Request Ancestor Package Info For Orphans) - will move on to part 3 soon.

instagibbs commented on src/node/txpackagetracker.cpp:
```diff
@@ -158,14 +164,26 @@ class TxPackageTracker::Impl {
+        } else {
+            // Even though this stores the orphan wtxid, is_wtxid=false because we will be requesting the parents via txid.
+            LogPrint(BCLog::TXPACKAGES, "\nPotential orphan resolution for %s from legacy peer=%d\n", wtxid.ToString(), nodeid);
+            orphan_request_tracker.ReceivedInv(nodeid, GenTxid::Txid(wtxid), is_preferred, reqtime);
```
It makes it very difficult to read, imo, and lilkely introduces bugs going forward, if not now.

Note @mzumsande that this logic lives in the orphan relay PR IIRC

mzumsande replied:
right - I'll copy the comment so that we can discuss there - this one can be resolved.

instagibbs commented:
I suspect this implementation, with an absurdly high ancestor chain limit being set, would violate the 100 txn limit for `getpackagetxns`: https://github.com/bitcoin/bips/blob/02ec218c7857ef60914e9a3d383b68caf987f70b/bip-0331.mediawiki#getpkgtxns

glozow commented:
> I suspect this implementation, with an absurdly high ancestor chain limit being set, would violate the 100 txn limit for getpackagetxns: https://github.com/bitcoin/bips/blob/02ec218c7857ef60914e9a3d383b68caf987f70b/bip-0331.mediawiki#getpkgtxns

Are you sure? We still require an `ancpkginfo` to have (non-configurable) `MAX_PACKAGE_COUNT` or fewer transactions, otherwise we exit "discarding packageinfo, too many transactions". But I can add a check that the list is <= `MAX_PKGTXNS_COUNT` prior to sending.

glozow committed:
fixup: make sure we never send getpkgtxns size > MAX_PKGTXNS_COUNT

glozow committed:
fixup: use single-sha256 for MSG_PKGTXNS combined hash

glozow commented:
Added check that outbound `getpkgtxns` is <= `MAX_PKGTXNS_COUNT`. Also changed the combined hash from double-sha256 to single-sha256 (which was changed in the BIP recently).

instagibbs commented:
Dropping note from offline discussion: probably worthwhile to just to advise users that setting ancestor count higher than 25 may end in weird behavior with chains longer than what package relay can support. I don't know of anyone who sets this value higher manually for production usage, and the value can be updated later easily as an implementation detail.

ariard commented on src/net_processing.cpp:
```diff
@@ -776,13 +801,18 @@ class PeerManagerImpl final : public PeerManager
+     *
+     * Upon receiving an announcement for a transaction, if it exists in this filter, do not
+     * download the txdata. Upon receiving a package info, if it contains a transaction in this
+     * filter, do not download the tx data.
```
I think it can say do not download "the whole package data", if this is referencing to the check L4415 in `src/net_processing.cpp` as pkg info itself is discarded.

glozow commented:
Not intended for merge. Not planning on rebasing right now since this should be built on top of cluster mempool and other WIP projects.

glozow commented:
Ok Drahty I'll take the hint :joy:.

Most of the code here is in master now - the Orphan Resolution Module is part of `TxDownloadManager` (#30110, #31397, etc.) and the `TxOrphanage` eviction changes (#31829, etc.) replace the token bucket protection stuff here. One difference is that we didn’t add the `TxRequestTracker orphan_request_tracker` for vanilla orphan resolution, but that is probably needed for ancestor package relay.

Parts 2 and 3 are relatively smaller, but still need the mempool changes. At this point a fresh implementation would be easier than a rebase here, so closing this.
